### PR TITLE
Optimize empty list constants

### PR DESCRIPTION
**Pull request by edolstra**
# Motivation

This avoids a `Value` allocation for empty list constants. During a `nix search nixpkgs`, about 82% of all thunked lists are empty, so this removes about 3 million `Value` allocations.

Performance comparison on `nix search github:NixOS/nixpkgs/e1fa12d4f6c6fe19ccb59cac54b5b3f25e160870 --no-eval-cache`:

```
maximum RSS:        median = 3845432.0000  mean = 3845432.0000  stddev =      0.0000  min = 3845432.0000  max = 3845432.0000  [rejected?, p=0.00000, Δ=-70084.00000±0.00000]
soft page faults:   median = 965395.0000  mean = 965394.6667  stddev =      1.1181  min = 965392.0000  max = 965396.0000  [rejected?, p=0.00000, Δ=-17929.77778±38.59610]
system CPU time:    median =      1.8029  mean =      1.7702  stddev =      0.0621  min =      1.6749  max =      1.8417  [rejected, p=0.00064, Δ=-0.12873±0.09905]
user CPU time:      median =     14.1022  mean =     14.0633  stddev =      0.1869  min =     13.8118  max =     14.3190  [not rejected, p=0.03006, Δ=-0.18248±0.24928]
elapsed time:       median =     15.8205  mean =     15.8618  stddev =      0.2312  min =     15.5033  max =     16.1670  [not rejected, p=0.00558, Δ=-0.28963±0.29434]
```

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/eval.hh [hunk 1]
```diff
@@ -305,6 +305,9 @@ public:
         return *errorBuilder;
     }
 
+    /* Empty list constant. */
+    Value vEmptyList;
```
Comment (roberth) [inline src/libexpr/eval.hh:309]: Maybe move this next to `emptyBindings` above?

**Review by roberth — COMMENTED** [review]
Looks ok.

Any reason not to do this for empty attrs too?
It'd be more consistent to have a singleton for both `[]` and `{}`.

**Comment by edolstra** [comment]
I tried the same for attrs but it didn't give a speedup. There are not nearly as many empty attrsets constants as empty lists.

**Comment by Ericson2314** [comment]
Well if there is no benefit but also no cost, it might still be good to do for consistency.

**Comment by edolstra** [comment]
> Well if there is no benefit but also no cost

There is always a cost. Every line of code is a maintenance burden.

**Pull request merged**